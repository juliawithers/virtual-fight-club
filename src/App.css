* {
  box-sizing: border-box;
}

.body {
  font-family: "Open Sans",
  sans-serif;
  font-size: 20px;
  color: #003333;
  text-align: center;
  background-color: #f9f1f1;
  height: 100%;
  position: relative;
}

html{
  height: 100%;
}

.container{
  min-height: 100vh;
 overflow: hidden;
 display: block;
 position: relative;
 padding-bottom: 70px; 
}

nav,
footer {
  padding: 1% 0;
  background: rgb(2,0,36);
  background: linear-gradient(#669999, #FF3333);
}

footer {
  text-align: center;
  position: absolute;
  bottom: 0;
  width: 100%;
}

nav {
  position: fixed;
  height: 50px;
  top: 0;
  left: 0;
  right: 0;
  z-index: 10;
}

nav > div {
  display: flex;
  justify-content: space-evenly;
}

.nav-link {
  text-decoration: none;
  color: #003333;
  padding: 5px;
}

.nav-link:hover{
  background-color: #f9f1f1;
  border-radius: 5px;
}

.link {
  text-decoration: none;
  color: #003333;
}

header{
  padding-top: 35px;
}

button{
  text-decoration: none;
  color: #003333;
  border: 2px solid #003333; 
  border-radius: 5px;
  padding: 4px;
  margin: 5px;
}

button:hover{
  cursor: pointer;
}

section {
  min-height: 200px;
  padding: 30px 15px;
}

p {
  margin: 0 auto;
  max-width: 650px;
}

p + p {
  margin-top: 15px;
}

input{
  padding: 5px;
}
