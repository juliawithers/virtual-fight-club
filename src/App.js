import React, { Component } from 'react';
import { Route, NavLink, withRouter } from 'react-router-dom'
import './App.css';
import Character from './Character/Character'
import Message from './Message/Message'
import HandleLanding from './HandleLanding/HandleLanding'
import AboutGame from './AboutGame/AboutGame'
import Fight from './Fight/Fight'
import context from './context'
import config from './config'

// auth tokens 
// style

class App extends Component {
  static contextType = context;
  constructor(props) {
    super(props);
    this.state = {
      characters: [],
      usernameErr: '',
      passwordErr: '',
      error: '',
      login: false,
      auth: '',
      username: '',
      user_id: '',
      character: {},
      attrpoints: '',
      level: '',
      opponent: {},
      handleLoginSubmit: () => { },
      submitUserInfo: () => { },
      updateCharacter: () => { },
      getCharacter: () => { },
      getCharactersList: () => { },
      createNewOpponent: () => { },
      createCharacter: () => { },
    }
  }

  componentDidMount() {
    fetch(config.API_CHARACTERS_ENDPOINT, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .then(characters => {
        this.setState({
          characters: characters,
        })
      })
      .catch(error => {
        console.log(error)
        this.setState({ error })
      })
  }

  createNewOpponent = (characters, userId) => {

    const opponents = characters.filter(player => player.user_id !== userId);

    const opponent = opponents[Math.floor(Math.random() * Math.floor(opponents.length))];  
    return opponent
  }

  handleLoginSubmit = (username, password) => {
    const object = {
      username: username,
      password: password
    }
    fetch(config.API_LOGIN_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify(object),
      headers: {
        'Content-Type': 'application/json',
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .then(data => {
        const opponent = this.createNewOpponent(this.state.characters, data.user.id)
        const character = this.state.characters.find(player => player.user_id === data.user.id)

        if (character === undefined) {
          this.setState({
            login: data.login,
            user_id: data.user.id,
            username: data.user.username,
            auth: data.user.auth,
            opponent: opponent,
            character: [],
            attrpoints: ''
          })
        } else {
          this.setState({
            login: data.login,
            user_id: data.user.id,
            username: data.user.username,
            auth: data.user.auth,
            opponent: opponent,
            character: character,
            attrpoints: character.attrpoints
          })  
        }
        
      })
      .catch(error => {
        this.setState({
          error: 'Password or Username is incorrect, please verify login information'
        })
      })
  }

  submitUserInfo = (object) => {
    const user_object = object;

    fetch(config.API_USERS_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify(user_object),
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .then(user => {
        this.setState({
          auth: user.auth,
          user_id: user.id,
          username: user.username,
          login: true
        })
      })
      .catch(error => this.setState({ error }))
  }
 
  createCharacter = (character) => {
    this.getCharactersList()
    fetch(config.API_CHARACTERS_ENDPOINT, {
      method: 'POST',
      body: JSON.stringify(character),
      headers: {
        'Content-Type': 'application/json',
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .then(character => {
        this.setState({
          character: character,
          attrpoints: Number(character.attrpoints)
        })
      })
      .catch(error => this.setState({ error }))
  }

  updateCharacter = (character, reason) => {
    fetch(config.API_CHARACTERS_ENDPOINT, {
      method: 'PATCH',
      body: JSON.stringify(character),
      headers: {
        'Content-Type': 'application/json',
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .catch(error => this.setState({ error }))
    if (reason === 'fight') {
      this.getCharacter(this.state.character.user_id)
      this.getCharactersList()
    } else if (reason === 'attributes') {
      this.getCharacter(this.state.character.user_id)
    }
  }

  getCharacter = (user_id) => { 
    fetch(config.API_CHARACTERS_ENDPOINT + `/${user_id}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .then(character => {
        this.setState({
          character: character
        })
      })
      .catch(error => this.setState({ error }))
  }

  getCharactersList = () => {
    fetch(config.API_CHARACTERS_ENDPOINT, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .then(characters => {
        this.setState({
          characters: characters
        })
      })
      .catch(error => this.setState({ error }))
  }

  deleteCharacter = (charId) => {
    const idObject = {id: charId};
    fetch(config.API_CHARACTERS_ENDPOINT, {
      method: 'DELETE',
      body: JSON.stringify(idObject),
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }
        return res.json()
      })
      .catch(error => this.setState({ error }))

  }

  deleteUser = (userId) => {
    const idObject = {id: userId}
    fetch(config.API_USERS_ENDPOINT, {
      method: 'DELETE',
      body: JSON.stringify(idObject),
      headers: {
        'Content-Type': 'application/json'
      }
    })
      .then(res => {
        if (!res.ok) {
          throw new Error(res.status)
        }

        this.setState({
          login: false
        })
        this.props.history.push('/')
        return res.json()
      })
      .catch(error => this.setState({ error }))
  }

  handleLogout = () => {
    this.setState({
      login: false,
      character: []
    })
    this.props.history.push('/')
  }

  createNavRoutes() {
    if (this.state.login === true) {
      return (
        <div>
          <NavLink
            className="nav-link"
            to="/aboutVFC">About</NavLink>
          <NavLink
            className="nav-link"
            to={`/auth/:${this.state.character.char_name}/character`}>Character</NavLink>
          <NavLink
            className="nav-link"
            to="/fight">FIGHT!</NavLink>
          <button className="logout" onClick={this.handleLogout}>
            Logout
          </button>
        </div>
      )
    } else {
      return (
        <div></div>
      )
    }
  }

  createMainRoutes() {
    return (
      <>
        <Route
          exact
          path="/"
          component={HandleLanding}
        />
        <Route
          exact
          path="/message"
          component={Message}
        />
        <Route
          exact
          path="/aboutVFC"
          component={AboutGame}
        />
        <Route
          exact
          path="/auth/:user_id/character"
          component={Character}
        />
        <Route
          exact
          path="/fight"
          component={Fight}
        />
      </>
    )
  }

  render() {
    
    const contextValue = {
      characters: this.state.characters,
      usernameErr: this.state.usernameErr,
      passwordErr: this.state.passwordErr,
      login: this.state.login,
      auth: this.state.auth,
      username: this.state.username,
      user_id: this.state.user_id,
      character: this.state.character,
      attrpoints: this.state.attrpoints,
      level: this.state.level,
      opponent: this.state.opponent,
      handleLoginSubmit: this.handleLoginSubmit,
      submitUserInfo: this.submitUserInfo,
      updateCharacter: this.updateCharacter,
      getCharacter: this.getCharacter,
      getCharactersList: this.getCharactersList,
      createNewOpponent: this.createNewOpponent,
      deleteCharacter: this.deleteCharacter,
      deleteUser: this.deleteUser,
      createCharacter: this.createCharacter,
      error: this.state.error
    }

    return (
      <context.Provider value={contextValue}>
        <body>
          <nav role="navigation">{this.createNavRoutes()}</nav>
          <main role="main">
            {this.createMainRoutes()}
          </main>
          <footer role="content-info">Virtual Fight Club</footer>
        </body>
      </context.Provider>
    )
  }
}


export default withRouter(App)