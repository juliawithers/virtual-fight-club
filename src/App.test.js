import React from 'react';
import { shallow, configure } from 'enzyme';
import toJson from 'enzyme-to-json';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import App from './App';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

describe(`App component`, () => {
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <BrowserRouter>
                <App />
            </BrowserRouter>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    })
    it('renders the UI as expected',() =>{
        const wrapper = shallow(<App/>)
        expect(toJson(wrapper)).toMatchSnapshot();
    })
})  