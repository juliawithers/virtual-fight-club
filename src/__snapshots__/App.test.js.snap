// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders the UI as expected 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
